pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go

```diff
@@ -133,35 +133,63 @@ func (self *Client) fetchVersion(authHeader string) error {
 	return nil
 }
 
+func (self *Client) apiStatus(resource string, auth string) (int, error) {
+	path := fmt.Sprintf("%s/api/%s/%s", self.host, self.version, resource)
+	resp, rErr := doRequest(self.client, "GET", path, auth, nil)
+	if rErr != nil {
+		return 0, errors.New(fmt.Sprintf("Couldn't connect to api: %s", rErr))
+	}
+
+	defer resp.Body.Close()
+	return resp.StatusCode, nil
+}
+
 func (self *Client) guessUserData(creds *Credentials) error {
 	// Do this explictly so that we know we have a valid response
 	// Kubernetes doesn't provide any way for a caller
 	// to find out who it is, so fill in the
 	// user as best we can.
+	creds.DisplayName = creds.UserName
+
 	err := self.fetchVersion(creds.GetHeader())
 	if err != nil {
 		return err
 	}
 
-	// If the API is open make sure we don't have any credentials
-	// so we aren't just saying yes to everything
-	resp, e := self.DoRequest("GET", "api", "", nil, nil)
-
-	// Treat connection errors as internal errors and invalid
-	// responses as auth errors
-	if e != nil {
-		return errors.New(fmt.Sprintf("Couldn't connect to api: %s", e))
+	// So we have a version, now if the API endpoint without any
+	// auth data returns a 401 our creds were good and we can log the user in.
+	var e error
+	var status int
+	status, e = self.apiStatus("", "")
+	success := status == 401
+
+	// if /api is configured to not require auth then
+	// the version check might not have properly verifed our basic auth.
+	// Getting a 200 or 403 means that might be the case.
+	// We need to determine if that is because no auth is configured
+	// Or if basic auth is broken.
+	if status == 200 || status == 403 {
+		// Send an empty basic header for namespaces that is guarenteed to be invalid.
+		// This should return a 200 if the api is open or a 401 if the
+		// api is protected.
+		status, e = self.apiStatus("namespaces", "Basic")
+
+		// Some versions of kubernetes return 403 instead of 401
+		// when presented with bad basic auth data. In those cases
+		// we need to refuse authentication.
+		// https://github.com/kubernetes/kubernetes/pull/41775
+		if status == 403 {
+			e = errors.New("This version of kubernetes is not supported. Turn off anonymous auth or upgrade.")
+		} else if status == 200 || status == 401 {
+			success = true
```

Comment:
There are multiple calls happening here, it's not all the same. The comments describe the steps i'll try to make them clearer. But it's not simple to follow.